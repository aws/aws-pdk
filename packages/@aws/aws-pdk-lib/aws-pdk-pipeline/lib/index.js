"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
__exportStar(require("./pdk-pipeline"), exports);
__exportStar(require("./pdk-pipeline-ts-project"), exports);
__exportStar(require("./pdk-pipeline-py-project"), exports);
__exportStar(require("./pdk-pipeline-java-project"), exports);
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiaW5kZXguanMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyIuLi9zcmMvaW5kZXgudHMiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6Ijs7Ozs7Ozs7Ozs7O0FBQUEsaURBQStCO0FBQy9CLDREQUEwQztBQUMxQyw0REFBMEM7QUFDMUMsOERBQTRDIiwic291cmNlc0NvbnRlbnQiOlsiZXhwb3J0ICogZnJvbSBcIi4vcGRrLXBpcGVsaW5lXCI7XG5leHBvcnQgKiBmcm9tIFwiLi9wZGstcGlwZWxpbmUtdHMtcHJvamVjdFwiO1xuZXhwb3J0ICogZnJvbSBcIi4vcGRrLXBpcGVsaW5lLXB5LXByb2plY3RcIjtcbmV4cG9ydCAqIGZyb20gXCIuL3Bkay1waXBlbGluZS1qYXZhLXByb2plY3RcIjtcbiJdfQ==