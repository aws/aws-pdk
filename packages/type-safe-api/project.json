{
  "name": "@aws/type-safe-api",
  "root": "packages/type-safe-api",
  "targets": {
    "build": {
      "inputs": [
        "default",
        "^default",
        "!{projectRoot}/.jsii",
        "!{projectRoot}/lib/**/*",
        "!{projectRoot}/dist/**/*",
        "!{projectRoot}/coverage/**/*",
        "!{projectRoot}/test-reports/**/*",
        "!{projectRoot}/docs/api/**/*"
      ],
      "outputs": [
        "{projectRoot}/.jsii",
        "{projectRoot}/lib",
        "{projectRoot}/dist",
        "{projectRoot}/coverage",
        "{projectRoot}/test-reports",
        "{projectRoot}/docs/api"
      ],
      "dependsOn": [
        "^build"
      ],
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen build",
        "cwd": "packages/type-safe-api"
      }
    },
    "default": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen default",
        "cwd": "packages/type-safe-api"
      }
    },
    "pre-compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen pre-compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "post-compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen post-compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "test": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen test",
        "cwd": "packages/type-safe-api"
      }
    },
    "package": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package",
        "cwd": "packages/type-safe-api"
      }
    },
    "test:watch": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen test:watch",
        "cwd": "packages/type-safe-api"
      }
    },
    "watch": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen watch",
        "cwd": "packages/type-safe-api"
      }
    },
    "eslint": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen eslint",
        "cwd": "packages/type-safe-api"
      }
    },
    "compat": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen compat",
        "cwd": "packages/type-safe-api"
      }
    },
    "package-all": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package-all",
        "cwd": "packages/type-safe-api"
      }
    },
    "package:java": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package:java",
        "cwd": "packages/type-safe-api"
      }
    },
    "package:python": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package:python",
        "cwd": "packages/type-safe-api"
      }
    },
    "eslint-staged": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen eslint-staged",
        "cwd": "packages/type-safe-api"
      }
    },
    "jest": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen jest",
        "cwd": "packages/type-safe-api"
      }
    },
    "docgen": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen docgen",
        "cwd": "packages/type-safe-api"
      }
    }
  },
  "implicitDependencies": [
    "smithy-async-transformer"
  ],
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \"pnpm exec projen\"."
}
