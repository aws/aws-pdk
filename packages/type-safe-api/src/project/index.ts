/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
export * from "./model/type-safe-api-model-build";
export * from "./model/type-safe-api-async-model-build";
export * from "./model/smithy/smithy-project-definition";
export * from "./model/openapi/open-api-project-definition";
export * from "./type-safe-api-project";
export * from "./model/type-safe-api-model-project";
export * from "./model/smithy/smithy-definition";
export * from "./model/openapi/open-api-definition";
export * from "./type-safe-websocket-api-project";
export * from "./model/type-safe-websocket-api-model-project";
export * from "./model/smithy/smithy-async-definition";
export * from "./model/openapi/open-api-async-definition";
export * from "./model/smithy/types";
export * from "./model/openapi/open-api-model-project";
export * from "./model/openapi/open-api-async-model-project";
export * from "./model/smithy/smithy-async-model-project";
export * from "./model/smithy/smithy-model-project";
export * from "./languages";
export * from "./types";
export * from "./typescript-project-options";
export * from "./python-project-options";
export * from "./java-project-options";
