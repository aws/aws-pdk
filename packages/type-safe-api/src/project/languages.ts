/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
/**
 * Supported languages for runtimes and infrastructure
 */
export enum Language {
  TYPESCRIPT = "typescript",
  PYTHON = "python",
  JAVA = "java",
}

/**
 * Versions of node
 */
export enum NodeVersion {
  NODE_18 = "NODE_18",
  NODE_20 = "NODE_20",
}

/**
 * Versions of java
 */
export enum JavaVersion {
  JAVA_8 = "JAVA_8",
  JAVA_11 = "JAVA_11",
  JAVA_17 = "JAVA_17",
  JAVA_21 = "JAVA_21",
}

/**
 * Versions of python
 */
export enum PythonVersion {
  PYTHON_3_11 = "PYTHON_3_11",
  PYTHON_3_12 = "PYTHON_3_12",
}

/**
 * Supported libraries for code generation
 */
export enum Library {
  TYPESCRIPT_REACT_QUERY_HOOKS = "typescript-react-query-hooks",
}

export enum WebSocketLibrary {
  TYPESCRIPT_WEBSOCKET_CLIENT = "typescript-websocket-client",
  TYPESCRIPT_WEBSOCKET_HOOKS = "typescript-websocket-hooks",
}

/**
 * Formats for documentation generation
 */
export enum DocumentationFormat {
  /**
   * HTML Documentation generated by redoc
   * @see https://github.com/Redocly/redoc
   */
  HTML_REDOC = "html_redoc",
  /**
   * OpenAPI Generator 'html2' documentation
   * @see https://github.com/OpenAPITools/openapi-generator/blob/master/docs/generators/html2.md
   */
  HTML2 = "html2",
  /**
   * OpenAPI Generator 'markdown' documentation
   * @see https://github.com/OpenAPITools/openapi-generator/blob/master/docs/generators/markdown.md
   */
  MARKDOWN = "markdown",
  /**
   * OpenAPI Generator 'plantuml' documentation
   * @see https://github.com/OpenAPITools/openapi-generator/blob/master/docs/generators/plantuml.md
   */
  PLANTUML = "plantuml",
}
