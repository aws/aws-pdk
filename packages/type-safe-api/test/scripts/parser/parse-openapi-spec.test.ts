/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import os from "os";
import * as path from "path";
import { exec } from "projen/lib/util";
import { withTmpDirSnapshot } from "../../project/snapshot-utils";

describe("Parse OpenAPI Spec Script Unit Tests", () => {
  it("Bundles Spec Into Single File", () => {
    expect(
      withTmpDirSnapshot(os.tmpdir(), (tmpDir) => {
        const specPath = "../../resources/specs/multi.yaml";
        const outputPath = path.join(
          path.relative(path.resolve(__dirname), tmpDir),
          ".api.json"
        );
        const command = `../../../scripts/type-safe-api/parser/parse-openapi-spec --spec-path ${specPath} --output-path ${outputPath}`;
        exec(command, {
          cwd: path.resolve(__dirname),
        });
      })
    ).toMatchSnapshot();
  });

  it("Injects @handler and @paginated traits", () => {
    expect(
      withTmpDirSnapshot(os.tmpdir(), (tmpDir) => {
        const specPath =
          "../../resources/smithy/simple-pagination/openapi.json";
        const smithyJsonModelPath =
          "../../resources/smithy/simple-pagination/model.json";
        const outputPath = path.join(
          path.relative(path.resolve(__dirname), tmpDir),
          ".api.json"
        );
        const command = `../../../scripts/type-safe-api/parser/parse-openapi-spec --spec-path ${specPath} --output-path ${outputPath} --smithy-json-path ${smithyJsonModelPath}`;
        exec(command, {
          cwd: path.resolve(__dirname),
        });
      })
    ).toMatchSnapshot();
  });

  it("Maps renamed @paginated traits for query and header parameters", () => {
    expect(
      withTmpDirSnapshot(os.tmpdir(), (tmpDir) => {
        const specPath =
          "../../resources/smithy/rename-pagination/openapi.json";
        const smithyJsonModelPath =
          "../../resources/smithy/rename-pagination/model.json";
        const outputPath = path.join(
          path.relative(path.resolve(__dirname), tmpDir),
          ".api.json"
        );
        const command = `../../../scripts/type-safe-api/parser/parse-openapi-spec --spec-path ${specPath} --output-path ${outputPath} --smithy-json-path ${smithyJsonModelPath}`;
        exec(command, {
          cwd: path.resolve(__dirname),
        });
      })
    ).toMatchSnapshot();
  });

  it("Throws for unsupported request parameter types", () => {
    withTmpDirSnapshot(os.tmpdir(), (tmpDir) => {
      const specPath = "../../resources/specs/invalid-request-parameters.yaml";
      const outputPath = path.join(
        path.relative(path.resolve(__dirname), tmpDir),
        ".api.json"
      );
      const command = `../../../scripts/type-safe-api/parser/parse-openapi-spec --spec-path ${specPath} --output-path ${outputPath}`;
      expect(() => {
        exec(command, {
          cwd: path.resolve(__dirname),
        });
      }).toThrow(
        /Request parameters must be of type number, integer, boolean, string or arrays of these. Found invalid parameters:\nget \/invalid\/parameters: objectQueryParam\nget \/invalid\/parameters: arrayOfObjects\nget \/invalid\/parameters: mixedTypes/
      );
    });
  });

  it("Permits parameter references (and circular references)", () => {
    expect(
      withTmpDirSnapshot(os.tmpdir(), (tmpDir) => {
        const specPath = "../../resources/specs/parameter-refs.yaml";
        const outputPath = path.join(
          path.relative(path.resolve(__dirname), tmpDir),
          ".api.json"
        );
        const command = `../../../scripts/type-safe-api/parser/parse-openapi-spec --spec-path ${specPath} --output-path ${outputPath}`;
        exec(command, {
          cwd: path.resolve(__dirname),
        });
      })
    ).toMatchSnapshot();
  });

  it("Injects @async, @connectHandler and @disconnectHandler traits", () => {
    expect(
      withTmpDirSnapshot(os.tmpdir(), (tmpDir) => {
        const specPath = "../../resources/smithy/async-recursive/openapi.json";
        const smithyJsonModelPath =
          "../../resources/smithy/async-recursive/model.json";
        const outputPath = path.join(
          path.relative(path.resolve(__dirname), tmpDir),
          ".api.json"
        );
        const command = `../../../scripts/type-safe-api/parser/parse-openapi-spec --spec-path ${specPath} --output-path ${outputPath} --smithy-json-path ${smithyJsonModelPath}`;
        exec(command, {
          cwd: path.resolve(__dirname),
        });
      })
    ).toMatchSnapshot();
  });
});
