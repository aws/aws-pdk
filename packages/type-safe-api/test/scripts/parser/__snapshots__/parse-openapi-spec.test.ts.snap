// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parse OpenAPI Spec Script Unit Tests Bundles Spec Into Single File 1`] = `
{
  ".api.json": {
    "components": {
      "schemas": {
        "ApiError": {
          "properties": {
            "errorMessage": {
              "type": "string",
            },
          },
          "required": [
            "errorMessage",
          ],
          "type": "object",
        },
        "TestRequest": {
          "properties": {
            "myInput": {
              "type": "number",
            },
          },
          "type": "object",
        },
        "TestResponse": {
          "properties": {
            "messages": {
              "items": {
                "properties": {
                  "id": {
                    "type": "integer",
                  },
                  "message": {
                    "type": "string",
                  },
                },
                "required": [
                  "id",
                ],
                "type": "object",
              },
              "type": "array",
            },
          },
          "required": [
            "messages",
          ],
          "type": "object",
        },
      },
    },
    "info": {
      "title": "Example API",
      "version": "1.0.0",
    },
    "openapi": "3.0.3",
    "paths": {
      "/operation/{pathParam}": {
        "post": {
          "operationId": "someTestOperation",
          "parameters": [
            {
              "in": "path",
              "name": "pathParam",
              "required": true,
              "schema": {
                "type": "string",
              },
            },
          ],
          "requestBody": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/TestRequest",
                },
              },
            },
            "required": true,
          },
          "responses": {
            "200": {
              "content": {
                "application/json": {
                  "schema": {
                    "$ref": "#/components/schemas/TestResponse",
                  },
                },
              },
              "description": "Successful response",
            },
            "400": {
              "content": {
                "application/json": {
                  "schema": {
                    "$ref": "#/components/schemas/ApiError",
                  },
                },
              },
              "description": "Error response",
            },
          },
        },
      },
    },
  },
}
`;

exports[`Parse OpenAPI Spec Script Unit Tests Injects @async, @connectHandler and @disconnectHandler traits 1`] = `
{
  ".api.json": {
    "components": {
      "schemas": {
        "Bar": {
          "properties": {
            "blip": {
              "type": "string",
            },
          },
          "type": "object",
        },
        "Foo": {
          "properties": {
            "bar": {
              "$ref": "#/components/schemas/Bar",
            },
            "baz": {
              "format": "int32",
              "type": "integer",
            },
          },
          "type": "object",
        },
        "Node": {
          "properties": {
            "left": {
              "$ref": "#/components/schemas/Node",
            },
            "right": {
              "$ref": "#/components/schemas/Node",
            },
            "value": {
              "type": "string",
            },
          },
          "required": [
            "value",
          ],
          "type": "object",
        },
        "RecursiveRequestContent": {
          "properties": {
            "root": {
              "$ref": "#/components/schemas/Node",
            },
          },
          "type": "object",
        },
        "SendNotificationRequestContent": {
          "properties": {
            "message": {
              "type": "string",
            },
            "title": {
              "type": "string",
            },
            "topic": {
              "type": "string",
            },
          },
          "required": [
            "message",
            "title",
            "topic",
          ],
          "type": "object",
        },
        "SubscribeToNotificationsRequestContent": {
          "properties": {
            "foo": {
              "$ref": "#/components/schemas/Foo",
            },
            "topic": {
              "type": "string",
            },
          },
          "required": [
            "topic",
          ],
          "type": "object",
        },
      },
    },
    "info": {
      "description": "A sample smithy websocket api",
      "title": "MyService",
      "version": "1.0",
    },
    "openapi": "3.0.2",
    "paths": {
      "/Recursive": {
        "post": {
          "operationId": "Recursive",
          "requestBody": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/RecursiveRequestContent",
                },
              },
            },
          },
          "responses": {
            "200": {
              "description": "Recursive 200 response",
            },
          },
          "x-async": {
            "direction": "bidirectional",
          },
          "x-smithy.api#http": {
            "code": 200,
            "method": "POST",
            "uri": "/Recursive",
          },
        },
      },
      "/SendNotification": {
        "post": {
          "operationId": "SendNotification",
          "requestBody": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/SendNotificationRequestContent",
                },
              },
            },
            "required": true,
          },
          "responses": {
            "200": {
              "description": "SendNotification 200 response",
            },
          },
          "x-async": {
            "direction": "server_to_client",
          },
          "x-smithy.api#http": {
            "code": 200,
            "method": "POST",
            "uri": "/SendNotification",
          },
        },
      },
      "/SubscribeToNotifications": {
        "post": {
          "operationId": "SubscribeToNotifications",
          "requestBody": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/SubscribeToNotificationsRequestContent",
                },
              },
            },
            "required": true,
          },
          "responses": {
            "200": {
              "description": "SubscribeToNotifications 200 response",
            },
          },
          "x-async": {
            "direction": "client_to_server",
          },
          "x-smithy.api#http": {
            "code": 200,
            "method": "POST",
            "uri": "/SubscribeToNotifications",
          },
        },
      },
    },
    "x-aws.protocols#restJson1": {},
    "x-com.aws#websocketJson": {},
    "x-connect-handler": {
      "language": "typescript",
    },
    "x-disconnect-handler": {
      "language": "typescript",
    },
    "x-smithy.api#documentation": "A sample smithy websocket api",
  },
}
`;

exports[`Parse OpenAPI Spec Script Unit Tests Injects @handler and @paginated traits 1`] = `
{
  ".api.json": {
    "components": {
      "schemas": {
        "BadRequestErrorResponseContent": {
          "description": "An error at the fault of the client sending invalid input",
          "properties": {
            "message": {
              "description": "Message with details about the error",
              "type": "string",
            },
          },
          "required": [
            "message",
          ],
          "type": "object",
        },
        "InternalFailureErrorResponseContent": {
          "description": "An internal failure at the fault of the server",
          "properties": {
            "message": {
              "description": "Message with details about the error",
              "type": "string",
            },
          },
          "required": [
            "message",
          ],
          "type": "object",
        },
        "ListHellosResponseContent": {
          "properties": {
            "messages": {
              "items": {
                "type": "string",
              },
              "type": "array",
            },
            "outNextToken": {
              "type": "string",
            },
          },
          "required": [
            "messages",
          ],
          "type": "object",
        },
        "NotAuthorizedErrorResponseContent": {
          "description": "An error due to the client not being authorized to access the resource",
          "properties": {
            "message": {
              "description": "Message with details about the error",
              "type": "string",
            },
          },
          "required": [
            "message",
          ],
          "type": "object",
        },
        "NotFoundErrorResponseContent": {
          "description": "An error due to the client attempting to access a missing resource",
          "properties": {
            "message": {
              "description": "Message with details about the error",
              "type": "string",
            },
          },
          "required": [
            "message",
          ],
          "type": "object",
        },
        "SayHelloResponseContent": {
          "properties": {
            "message": {
              "type": "string",
            },
          },
          "required": [
            "message",
          ],
          "type": "object",
        },
      },
    },
    "info": {
      "description": "A sample smithy api",
      "title": "MyService",
      "version": "1.0",
    },
    "openapi": "3.0.2",
    "paths": {
      "/hello": {
        "get": {
          "operationId": "SayHello",
          "parameters": [
            {
              "in": "query",
              "name": "name",
              "required": true,
              "schema": {
                "type": "string",
              },
            },
          ],
          "responses": {
            "200": {
              "content": {
                "application/json": {
                  "schema": {
                    "$ref": "#/components/schemas/SayHelloResponseContent",
                  },
                },
              },
              "description": "SayHello 200 response",
            },
            "400": {
              "content": {
                "application/json": {
                  "schema": {
                    "$ref": "#/components/schemas/BadRequestErrorResponseContent",
                  },
                },
              },
              "description": "BadRequestError 400 response",
            },
            "403": {
              "content": {
                "application/json": {
                  "schema": {
                    "$ref": "#/components/schemas/NotAuthorizedErrorResponseContent",
                  },
                },
              },
              "description": "NotAuthorizedError 403 response",
            },
            "404": {
              "content": {
                "application/json": {
                  "schema": {
                    "$ref": "#/components/schemas/NotFoundErrorResponseContent",
                  },
                },
              },
              "description": "NotFoundError 404 response",
            },
            "500": {
              "content": {
                "application/json": {
                  "schema": {
                    "$ref": "#/components/schemas/InternalFailureErrorResponseContent",
                  },
                },
              },
              "description": "InternalFailureError 500 response",
            },
          },
          "x-handler": {
            "language": "typescript",
          },
          "x-smithy.api#http": {
            "method": "GET",
            "uri": "/hello",
          },
          "x-smithy.api#readonly": {},
        },
      },
      "/hello/{foo}": {
        "get": {
          "operationId": "ListHellos",
          "parameters": [
            {
              "in": "path",
              "name": "foo",
              "required": true,
              "schema": {
                "type": "string",
              },
            },
            {
              "in": "query",
              "name": "nextToken",
              "schema": {
                "type": "string",
              },
            },
          ],
          "responses": {
            "200": {
              "content": {
                "application/json": {
                  "schema": {
                    "$ref": "#/components/schemas/ListHellosResponseContent",
                  },
                },
              },
              "description": "ListHellos 200 response",
            },
            "400": {
              "content": {
                "application/json": {
                  "schema": {
                    "$ref": "#/components/schemas/BadRequestErrorResponseContent",
                  },
                },
              },
              "description": "BadRequestError 400 response",
            },
            "403": {
              "content": {
                "application/json": {
                  "schema": {
                    "$ref": "#/components/schemas/NotAuthorizedErrorResponseContent",
                  },
                },
              },
              "description": "NotAuthorizedError 403 response",
            },
            "404": {
              "content": {
                "application/json": {
                  "schema": {
                    "$ref": "#/components/schemas/NotFoundErrorResponseContent",
                  },
                },
              },
              "description": "NotFoundError 404 response",
            },
            "500": {
              "content": {
                "application/json": {
                  "schema": {
                    "$ref": "#/components/schemas/InternalFailureErrorResponseContent",
                  },
                },
              },
              "description": "InternalFailureError 500 response",
            },
          },
          "x-paginated": {
            "inputToken": "nextToken",
            "items": "messages",
            "outputToken": "outNextToken",
          },
          "x-smithy.api#http": {
            "method": "GET",
            "uri": "/hello/{foo}",
          },
          "x-smithy.api#readonly": {},
        },
      },
    },
    "x-aws.protocols#restJson1": {},
    "x-smithy.api#documentation": "A sample smithy api",
  },
}
`;

exports[`Parse OpenAPI Spec Script Unit Tests Maps renamed @paginated traits for query and header parameters 1`] = `
{
  ".api.json": {
    "components": {
      "schemas": {
        "BadRequestErrorResponseContent": {
          "description": "An error at the fault of the client sending invalid input",
          "properties": {
            "message": {
              "description": "Message with details about the error",
              "type": "string",
            },
          },
          "required": [
            "message",
          ],
          "type": "object",
        },
        "InternalFailureErrorResponseContent": {
          "description": "An internal failure at the fault of the server",
          "properties": {
            "message": {
              "description": "Message with details about the error",
              "type": "string",
            },
          },
          "required": [
            "message",
          ],
          "type": "object",
        },
        "NotAuthorizedErrorResponseContent": {
          "description": "An error due to the client not being authorized to access the resource",
          "properties": {
            "message": {
              "description": "Message with details about the error",
              "type": "string",
            },
          },
          "required": [
            "message",
          ],
          "type": "object",
        },
        "NotFoundErrorResponseContent": {
          "description": "An error due to the client attempting to access a missing resource",
          "properties": {
            "message": {
              "description": "Message with details about the error",
              "type": "string",
            },
          },
          "required": [
            "message",
          ],
          "type": "object",
        },
        "SayHelloResponseContent": {
          "properties": {
            "messages": {
              "items": {
                "type": "string",
              },
              "type": "array",
            },
            "outToken": {
              "type": "string",
            },
          },
          "required": [
            "messages",
          ],
          "type": "object",
        },
      },
    },
    "info": {
      "description": "A sample smithy api",
      "title": "MyService",
      "version": "1.0",
    },
    "openapi": "3.0.2",
    "paths": {
      "/hello": {
        "get": {
          "operationId": "SayHello",
          "parameters": [
            {
              "in": "query",
              "name": "name",
              "required": true,
              "schema": {
                "type": "string",
              },
            },
            {
              "in": "query",
              "name": "myInToken",
              "schema": {
                "type": "string",
              },
            },
            {
              "in": "header",
              "name": "x-page-size",
              "schema": {
                "format": "int32",
                "type": "integer",
              },
            },
          ],
          "responses": {
            "200": {
              "content": {
                "application/json": {
                  "schema": {
                    "$ref": "#/components/schemas/SayHelloResponseContent",
                  },
                },
              },
              "description": "SayHello 200 response",
            },
            "400": {
              "content": {
                "application/json": {
                  "schema": {
                    "$ref": "#/components/schemas/BadRequestErrorResponseContent",
                  },
                },
              },
              "description": "BadRequestError 400 response",
            },
            "403": {
              "content": {
                "application/json": {
                  "schema": {
                    "$ref": "#/components/schemas/NotAuthorizedErrorResponseContent",
                  },
                },
              },
              "description": "NotAuthorizedError 403 response",
            },
            "404": {
              "content": {
                "application/json": {
                  "schema": {
                    "$ref": "#/components/schemas/NotFoundErrorResponseContent",
                  },
                },
              },
              "description": "NotFoundError 404 response",
            },
            "500": {
              "content": {
                "application/json": {
                  "schema": {
                    "$ref": "#/components/schemas/InternalFailureErrorResponseContent",
                  },
                },
              },
              "description": "InternalFailureError 500 response",
            },
          },
          "x-handler": {
            "language": "typescript",
          },
          "x-paginated": {
            "inputToken": "myInToken",
            "items": "messages",
            "outputToken": "outToken",
            "pageSize": "x-page-size",
          },
          "x-smithy.api#http": {
            "method": "GET",
            "uri": "/hello",
          },
          "x-smithy.api#readonly": {},
        },
      },
    },
    "x-aws.protocols#restJson1": {},
    "x-smithy.api#documentation": "A sample smithy api",
  },
}
`;

exports[`Parse OpenAPI Spec Script Unit Tests Permits parameter references (and circular references) 1`] = `
{
  ".api.json": {
    "components": {
      "parameters": {
        "HelloId": {
          "in": "query",
          "name": "id",
          "required": false,
          "schema": {
            "$ref": "#/components/schemas/HelloId",
          },
        },
      },
      "schemas": {
        "HelloId": {
          "type": "string",
        },
        "HelloResponse": {
          "properties": {
            "id": {
              "$ref": "#/components/schemas/HelloId",
            },
            "message": {
              "$ref": "#/components/schemas/HelloResponse",
            },
          },
          "required": [
            "id",
          ],
          "type": "object",
        },
      },
    },
    "info": {
      "title": "Example API",
      "version": "1.0.0",
    },
    "openapi": "3.0.3",
    "paths": {
      "/hello": {
        "get": {
          "operationId": "sayHello",
          "parameters": [
            {
              "$ref": "#/components/parameters/HelloId",
            },
          ],
          "responses": {
            "200": {
              "content": {
                "application/json": {
                  "schema": {
                    "$ref": "#/components/schemas/HelloResponse",
                  },
                },
              },
              "description": "Successful response",
            },
          },
          "x-handler": {
            "language": "typescript",
          },
        },
      },
    },
  },
}
`;
